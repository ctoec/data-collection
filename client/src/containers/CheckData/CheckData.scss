.CheckData__content {
  margin-bottom: 5rem;
}

.CheckData__button-container {
  .margin-bottom-0 {
    // USWDS grid container overrides this
    margin-bottom: 0 !important;
  }
  .fixed-buttons {
    box-shadow: 0 -2px 4px 0px rgba(0, 0, 0, 0.25) !important;
    clip-path: inset(-1rem 0 0 0);
    background-color: #fff;
    padding: 1rem;
  }
}

.no-wrap {
  width: max-content;
}
