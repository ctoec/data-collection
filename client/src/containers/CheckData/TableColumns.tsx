import React from 'react';
import idx from 'idx';
import { Link } from 'react-router-dom';
import { Column, Button } from '@ctoec/component-library';
import { Child } from '../../shared/models';

const tableColumnClassName = 'text-pre text-center font-body-2xs';
export const tableColumns: () => Column<Child>[] = () => {
  return [
    {
      className: tableColumnClassName,
      name: 'Name',
      sort: (row) => row.lastName,
      cell: ({ row }) => {
        return (
          <td>
            <Link
              to={{
                pathname: `/edit-record/${row.id}`,
              }}
            >
              <Button
                className="text-no-wrap font-body-2xs"
                appearance="unstyled"
                text={`${row.firstName} ${row.lastName}`}
              />
            </Link>
          </td>
        );
      },
    },
    {
      className: tableColumnClassName,
      name: 'Birth Date',
      cell: ({ row }) => (
        <td>{row.birthdate ? row.birthdate.format('MM/DD/YYYY') : ''}</td>
      ),
    },
    {
      className: tableColumnClassName,
      name: 'Age group',
      cell: ({ row }) => <td>{idx(row, (_) => _.enrollments[0].ageGroup)}</td>,
    },
    {
      className: tableColumnClassName,
      name: 'Funding type',
      cell: ({ row }) => (
        <td>
          {idx(row, (_) => _.enrollments[0].fundings[0].fundingSpace.source)}
        </td>
      ),
    },
    {
      className: tableColumnClassName,
      name: 'Contract space',
      cell: ({ row }) => (
        <td>
          {idx(row, (_) => _.enrollments[0].fundings[0].fundingSpace.time)}
        </td>
      ),
    },
    {
      className: tableColumnClassName,
      name: 'Site',
      cell: ({ row }) => <td>{idx(row, (_) => _.enrollments[0].site.name)}</td>,
    },
    {
      className: tableColumnClassName,
      name: 'Enrollment date',
      cell: ({ row }) => (
        <td>{idx(row, (_) => _.enrollments[0].entry.format('MM/DD/YYYY'))}</td>
      ),
    },
  ];
};
