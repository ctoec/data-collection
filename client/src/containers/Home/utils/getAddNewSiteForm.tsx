import { TextInput } from '@ctoec/component-library';
import React from 'react';
import { AddSiteRequest } from '../../../shared/payloads/AddSiteRequest';

/**
 * Function to expand a new set of fields for a user to add a new
 * site in the revision request form. The form tracks which site
 * to update among the state list sites by the index in the list.
 * @param addRequest
 * @param idx
 * @param setAddRequests
 */
export const getAddNewSiteForm = (
  addRequest: AddSiteRequest,
  idx: number,
  setAddRequests: React.Dispatch<React.SetStateAction<AddSiteRequest[]>>
) => {
  return (
    <>
      <p className="text-bold">New site information</p>
      <TextInput
        label="New site name"
        id="new-site-name"
        type="input"
        defaultValue={addRequest.siteName}
        onChange={(e) => {
          const change = e.target.value;
          setAddRequests((o: AddSiteRequest[]) => {
            o[idx].siteName = change;
            return o;
          });
          return change;
        }}
      />
      <TextInput
        label="License ID"
        id="new-site-license"
        type="input"
        defaultValue={addRequest.licenseId}
        onChange={(e) => {
          const change = e.target.value;
          setAddRequests((o: AddSiteRequest[]) => {
            o[idx].licenseId = change;
            return o;
          });
          return change;
        }}
      />
      <TextInput
        label="NAEYC ID"
        id="new-site-naeyc"
        type="input"
        defaultValue={addRequest.naeycId}
        onChange={(e) => {
          const change = e.target.value;
          setAddRequests((o: AddSiteRequest[]) => {
            o[idx].naeycId = change;
            return o;
          });
          return change;
        }}
      />
      <TextInput
        className="margin-bottom-4"
        label="Registry ID"
        id="new-site-registry"
        type="input"
        defaultValue={addRequest.registryId}
        onChange={(e) => {
          const change = e.target.value;
          setAddRequests((o: AddSiteRequest[]) => {
            o[idx].registryId = change;
            return o;
          });
          return change;
        }}
      />
    </>
  );
};
