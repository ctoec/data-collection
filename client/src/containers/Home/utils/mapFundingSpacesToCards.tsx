import React from 'react';
import { Card } from '@ctoec/component-library';
import { NestedFundingSpaces } from '../../../shared/payloads/NestedFundingSpaces';
import { AgeGroup, FundingSource } from '../../../shared/models';
import { getStrippedFundingSourceName } from './getFundingSpaceDisplayName';

// Map each calculated funding space distribution into a card
// element that we can format for display
export const mapFundingSpacesToCards = (nestedSpaces: NestedFundingSpaces) =>
  Object.keys(nestedSpaces).map((source) => (
    <div className="desktop:grid-col-4 three-column-card">
      <Card>
        <div className="padding-0">
          <h3 className="margin-top-0">
            {getStrippedFundingSourceName(source as FundingSource)}
          </h3>
          {Object.keys(nestedSpaces[source as FundingSource]).map(
            (ageGroup) => (
              <>
                <div>
                  <p className="text-bold">{ageGroup}</p>
                </div>
                <>
                  {nestedSpaces[source as FundingSource][
                    ageGroup as AgeGroup
                  ].map((fundingSpace) => {
                    // Account for fundings that have negative capacities stored
                    // Also hide capacities for SR fundings, since we're less
                    // confident on those and they change a lot
                    let spaceNumbers = `${fundingSpace.filled}`;
                    if (
                      fundingSpace.capacity !== -1 &&
                      !source.includes('School Readiness')
                    )
                      spaceNumbers += `/${fundingSpace.capacity} `;
                    else spaceNumbers += ` `;
                    return (
                      <p className="text-base-darker padding-0 margin-0">
                        <b>{spaceNumbers}</b>
                        {` ${fundingSpace.time}`}
                      </p>
                    );
                  })}
                </>
              </>
            )
          )}
        </div>
      </Card>
    </div>
  ));
