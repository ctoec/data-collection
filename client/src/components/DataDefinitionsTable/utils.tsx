import React from 'react';
import cx from 'classnames';
import { Tag } from '@ctoec/component-library';
import {
  TEMPLATE_REQUIREMENT_LEVELS,
  TEMPLATE_SECTIONS,
} from '../../shared/constants';
import { ColumnMetadata, Organization, Site } from '../../shared/models';
import { DATA_DEF_COLUMN_NAMES } from './TableColumns';

export const getRequiredTag = (requirementLevel: string) => (
  <Tag
    text={requirementLevel}
    className={cx(
      'required-tag',
      {
        'required-tag--required':
          requirementLevel === TEMPLATE_REQUIREMENT_LEVELS.REQUIRED,
      },
      {
        'required-tag--conditional':
          requirementLevel === TEMPLATE_REQUIREMENT_LEVELS.CONDITIONAL,
      }
    )}
  />
);

const FIRST_REPORTING_PERIOD_ALERT_NAME = 'firstReportingPeriodAlert';
export const FIRST_REPORTING_PERIOD_ALERT_ROW = {
  propertyName: FIRST_REPORTING_PERIOD_ALERT_NAME,
  section: TEMPLATE_SECTIONS.ENROLLMENT_FUNDING,
} as ColumnMetadata;

export const isFirstReportingPeriodRow = (row: ColumnMetadata) =>
  !!row.propertyName.match(/firstreporting/i);

export const isFirstReportingPeriodAlertRow = (row: ColumnMetadata) =>
  row.propertyName === FIRST_REPORTING_PERIOD_ALERT_NAME;

export const getMarkdownStyledFormatOptionsList = (formatString: string) => {
  const match = formatString.match(/^(One of:)/);
  if (match) {
    // Strip the leading 'One of:'
    return (
      formatString
        .replace(match[0], '')
        // then split by expected ', ' string
        .split(', ')
        // prepend markdown list character
        .map((li) => `- ${li}`)
        // then re-combine into two-space + new-line separated markdown list
        .join('  \n')
    );
  }

  return formatString;
};

export type EnhancedColumnMetadata = ColumnMetadata & {
  columnFormatters?: {
    [key in DATA_DEF_COLUMN_NAMES]?: (
      row: EnhancedColumnMetadata
    ) => React.ReactElement;
  };
};

export const getSiteFormatters = (sites: Site[]) => ({
  [DATA_DEF_COLUMN_NAMES.format]: () => (
    <td>
      Text, one of:
      <div className="margin-top-1">
        <ul>
          {sites.map((s) => (
            <li key={s.siteName}>{s.siteName}</li>
          ))}
        </ul>
      </div>
    </td>
  ),
});

export const getProviderFormatters = (organizations?: Organization[]) => ({
  [DATA_DEF_COLUMN_NAMES.format]: () => {
    if (!organizations?.length) {
      return <></>;
    }
    if (organizations?.length === 1) {
      return (
        <td>
          Text:
          <div className="margin-top-1">{organizations[0].providerName}</div>
        </td>
      );
    }
    return (
      <td>
        Text:
        <div className="margin-top-1">
          <ul>
            {organizations?.map((o) => (
              <li key={o.providerName}>{o.providerName}</li>
            ))}
          </ul>
        </div>
      </td>
    );
  },
});
