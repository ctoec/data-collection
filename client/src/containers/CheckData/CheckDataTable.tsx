import React from 'react';
import { tableColumns } from './TableColumns';
import { Child } from '../../shared/models';
import { Table } from '@ctoec/component-library';

type CheckDataTableProps = {
    reportId: number;
    reportData: Child[];
}

export const CheckDataTable: React.FC<CheckDataTableProps> = (reportId, reportData) => {
    console.log(reportData);
    return (
        <Table<Child>
            id="enrollment-report-table"
            rowKey={(row) => row.id}
            data={reportData}
            columns={tableColumns}
        />
    );
}