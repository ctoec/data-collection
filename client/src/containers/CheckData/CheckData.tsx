import React, { useContext, useState, useEffect } from 'react';
import { useLocation } from 'react-router-dom';
import queryString from 'query-string';
import MaterialTable from 'material-table';
import {
  tableIcons,
  DeleteOutline,
  oecFontFamily,
} from '../../utils/materialTable';
import AuthenticationContext from '../../contexts/AuthenticationContext/AuthenticationContext';
import { FlattenedEnrollment } from 'models';
import { apiGet } from '../../utils/api';

const CheckData: React.FC = () => {
  const { accessToken } = useContext(AuthenticationContext);
  const [reportData, setReportData] = useState<FlattenedEnrollment[]>([]);
  const reportId = parseInt(
    (queryString.parse(useLocation().search).reportId as string) || ''
  );

  useEffect(() => {
    apiGet(`enrollment-reports/${reportId}`, { accessToken }).then((report) => {
      if (report && report.enrollments) {
        setReportData(report.enrollments);
      }
    });
  }, [reportId, accessToken]);

  return (
    <MaterialTable
      icons={tableIcons}
      title="Example Data Table"
      data={reportData}
      columns={[
        {
          title: 'Name',
          field: 'name',
          validate: (rowData) => rowData.name !== '',
        },
      ]}
      editable={{
        onRowAdd,
        onRowUpdate,
        onRowDelete,
      }}
      actions={[
        {
          tooltip: 'Remove All Selected Users',
          icon: () => <DeleteOutline />,
          onClick: (evt, data) => alert('You want to delete rows?'),
        },
      ]}
      options={{
        rowStyle: {
          fontFamily: oecFontFamily,
        },
        headerStyle: {
          fontFamily: oecFontFamily,
        },
        selection: true,
        showTitle: false,
      }}
    />
  );
};

export default CheckData;

///////////////////////////////////////////////////

//  TODO: Capture and store new data entered to data table
function onRowAdd(newData: any) {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      resolve();
    }, 1000);
  });
}

//  TODO: Capture and store data table updates
function onRowUpdate(newData: any, oldData: any) {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      resolve();
    }, 1000);
  });
}

//  TODO: Capture and apply deletions made to data table
function onRowDelete(oldData: any) {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      resolve();
    }, 1000);
  });
}
