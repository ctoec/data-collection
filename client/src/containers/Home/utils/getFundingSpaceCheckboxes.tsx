import { Checkbox } from '@ctoec/component-library';
import React from 'react';
import { FUNDING_SOURCE_TIMES } from '../../../shared/constants';
import { AgeGroup } from '../../../shared/models';
import { ChangeFundingSpaceRequest } from '../../../shared/payloads/ChangeFundingSpaceRequest';
import { getStrippedFundingSourceName } from '../../../utils/getFundingSpaceDisplayName';

/**
 * Function that creates a series of checkboxes broken out by age
 * group at the bottom of the revision request form. Funding space
 * types that the user and their org already have access to are
 * pre-checked.
 * @param userFundingSpaces
 * @param setUserFundingSpaces
 */
export const getFundingSpaceCheckboxes = (
  userFundingSpaces: Partial<ChangeFundingSpaceRequest>[],
  setUserFundingSpaces: React.Dispatch<
    React.SetStateAction<Partial<ChangeFundingSpaceRequest>[]>
  >
) => {
  return Object.values(AgeGroup).map((ag) => (
    <>
      <p className="text-bold">{ag}</p>
      {FUNDING_SOURCE_TIMES.map((fst) => {
        if (!fst.ageGroupLimitations || fst.ageGroupLimitations.includes(ag)) {
          return fst.fundingSources.map((source) => {
            const sourceName = getStrippedFundingSourceName(source);
            return fst.fundingTimes.map((time) => {
              const rep = ag + ' - ' + sourceName + ' - ' + time.value;
              return (
                <Checkbox
                  id={`funding-space-check-${ag}-${
                    sourceName + ' - ' + time.value
                  }`}
                  text={sourceName + ' - ' + time.value}
                  defaultChecked={
                    userFundingSpaces.find((elt) => elt.fundingSpace === rep)
                      ?.shouldHave
                      ? true
                      : false
                  }
                  onChange={(e) => {
                    const checked = e.target.checked;
                    const rep = ag + ' - ' + sourceName + ' - ' + time.value;
                    setUserFundingSpaces(
                      (o: Partial<ChangeFundingSpaceRequest>[]) => {
                        const match = o.find(
                          (elt: Partial<ChangeFundingSpaceRequest>) =>
                            elt.fundingSpace === rep
                        );
                        if (match) match.shouldHave = checked;
                        return o;
                      }
                    );
                    return checked;
                  }}
                />
              );
            });
          });
        }
      })}
    </>
  ));
};
