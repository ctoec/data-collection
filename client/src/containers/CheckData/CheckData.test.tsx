import React from 'react';
import { waitFor } from '@testing-library/react';
import { Child, FundingSource } from '../../shared/models';
import CheckData from './CheckData';
import { accessibilityTestHelper, snapshotTestHelper } from '../../testHelpers';

// mock react-router-dom useParams
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useParams: () => ({ reportId: '1' }),
}));

// mock API response to generate table data
jest.mock('../../utils/api');
import * as api from '../../utils/api';
const apiMock = api as jest.Mocked<typeof api>;
const child = {
  id: '00000000-0000-0000-0000-0000000000000',
  firstName: 'First',
  lastName: 'Last',
  enrollments: [
    {
      site: { siteName: 'Site' },
      fundings: [
        {
          fundingSpace: { source: FundingSource.CDC },
        },
      ],
    },
  ],
} as Child;

describe('CheckData', () => {
  beforeAll(() =>
    apiMock.apiGet.mockReturnValue(new Promise((resolve) => resolve([child])))
  );

  const helperOpts = {
    wrapInRouter: true,
    before: () => waitFor(() => expect(apiMock.apiGet).toBeCalled()),
  };
  snapshotTestHelper(<CheckData />, helperOpts);
  accessibilityTestHelper(<CheckData />, helperOpts);

  afterAll(() => jest.clearAllMocks());
});
