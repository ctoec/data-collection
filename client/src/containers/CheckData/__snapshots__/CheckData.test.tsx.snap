// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckData matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="CheckData__content margin-4 grid-container"
  >
    <div>
      <button
        class="usa-button usa-button--unstyled margin-bottom-2 text-bold"
        type="button"
      >
        <span
          class="oec-text-with-icon"
        >
          <svg
            class="oec-text-with-icon__icon oec-text-with-icon__icon--direction-left oec-text-with-icon__icon--side-left"
          >
            arrowRight.svg
          </svg>
          Back
        </span>
      </button>
      <h1>
        Check data for 1 child
      </h1>
      <p>
        Make sure all of your data was uploaded correctly. 
      </p>
      <p>
        If everything looks good, submit to OEC.
      </p>
      <a
        href="/create-record"
      >
        Add a record
      </a>
      <div
        class="scrollbar-container ps"
      >
        <table
          class="oec-table"
          id="enrollment-report-table"
        >
          <thead>
            <tr>
              <th
                aria-sort="none"
                class="oec-table__column-header oec-sortable oec-unsorted text-pre text-center font-body-2xs"
                role="columnheader"
                scope="col"
              >
                <button
                  aria-label="Sort table by Name in ascending order"
                  class="oec-table__column-title usa-button--unstyled width-full"
                >
                  Name
                  <div
                    class="oec-table__sort-controls"
                  />
                </button>
              </th>
              <th
                aria-sort="none"
                class="oec-table__column-header oec-unsorted text-pre text-center font-body-2xs"
                role="columnheader"
                scope="col"
              >
                <span
                  class="oec-table__column-title"
                >
                  Birth Date
                </span>
              </th>
              <th
                aria-sort="none"
                class="oec-table__column-header oec-unsorted text-pre text-center font-body-2xs"
                role="columnheader"
                scope="col"
              >
                <span
                  class="oec-table__column-title"
                >
                  Age group
                </span>
              </th>
              <th
                aria-sort="none"
                class="oec-table__column-header oec-unsorted text-pre text-center font-body-2xs"
                role="columnheader"
                scope="col"
              >
                <span
                  class="oec-table__column-title"
                >
                  Funding type
                </span>
              </th>
              <th
                aria-sort="none"
                class="oec-table__column-header oec-unsorted text-pre text-center font-body-2xs"
                role="columnheader"
                scope="col"
              >
                <span
                  class="oec-table__column-title"
                >
                  Contract space
                </span>
              </th>
              <th
                aria-sort="none"
                class="oec-table__column-header oec-unsorted text-pre text-center font-body-2xs"
                role="columnheader"
                scope="col"
              >
                <span
                  class="oec-table__column-title"
                >
                  Site
                </span>
              </th>
              <th
                aria-sort="none"
                class="oec-table__column-header oec-unsorted text-pre text-center font-body-2xs"
                role="columnheader"
                scope="col"
              >
                <span
                  class="oec-table__column-title"
                >
                  Enrollment date
                </span>
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <th
                class="font-body-2xs"
                scope="row"
              >
                <a
                  href="/edit-record/00000000-0000-0000-0000-0000000000000"
                >
                  <button
                    class="usa-button usa-button--unstyled text-no-wrap font-body-2xs"
                    type="button"
                  >
                    First Last
                  </button>
                </a>
              </th>
              <td
                class="font-body-2xs"
              />
              <td
                class="font-body-2xs"
              />
              <td
                class="font-body-2xs"
              >
                CDC - Child Day Care
              </td>
              <td
                class="font-body-2xs"
              />
              <td
                class="font-body-2xs"
              >
                Site
              </td>
              <td
                class="font-body-2xs"
              />
            </tr>
          </tbody>
        </table>
        <div
          class="ps__rail-x"
          style="left: 0px; top: 0px;"
        >
          <div
            class="ps__thumb-x"
            style="left: 0px; width: 0px;"
            tabindex="0"
          />
        </div>
        <div
          class="ps__rail-y"
          style="top: 0px; left: 0px;"
        >
          <div
            class="ps__thumb-y"
            style="top: 0px; height: 0px;"
            tabindex="0"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    class="fixed-bottom-bar"
  >
    <div
      class="shadow-box"
    >
      <div
        class="grid-container"
      >
        <a
          class="usa-button usa-button--outline"
          href="/upload"
        >
          Back to upload
        </a>
        <a
          class="usa-button"
          href="/success/1"
        >
          Send to OEC
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
